File: src/main/java/alfheim/client/core/handler/PacketHandlerClient.java
package alfheim.client.core.handler;

import alexsocol.asjlib.ASJUtilities;
import alexsocol.asjlib.extendables.ItemContainingTileEntity;
import alfheim.api.AlfheimAPI;
import alfheim.api.entity.EnumRace;
import alfheim.api.spell.SpellBase.SpellCastResult;
import alfheim.client.core.handler.CardinalSystemClient.SpellCastingSystemClient;
import alfheim.client.core.handler.CardinalSystemClient.TimeStopSystemClient;
import alfheim.client.core.proxy.ClientProxy;
import alfheim.client.render.world.SpellEffectHandlerClient;
import alfheim.client.render.world.SpellEffectHandlerClient.Spells;
import alfheim.common.core.handler.CardinalSystem.HotSpellsSystem;
import alfheim.common.core.handler.CardinalSystem.PartySystem;
import alfheim.common.core.handler.CardinalSystem.PartySystem.Party;
import alfheim.common.core.handler.CardinalSystem.PartySystem.Party.PartyStatus;
import alfheim.common.core.registry.AlfheimRegistry;
import alfheim.common.core.util.AlfheimConfig;
import alfheim.common.entity.Flight;
import alfheim.common.network.Message1d;
import alfheim.common.network.MessageHotSpellC;
import alfheim.common.network.MessageParticles;
import alfheim.common.network.MessageParty;
import alfheim.common.network.MessageTileItem;
import alfheim.common.network.MessageTimeStop;
import alfheim.common.network.Message1d.m1d;
import alfheim.common.network.Message2d;
import alfheim.common.network.Message2d.m2d;
import alfheim.common.network.Message3d.m3d;
import alfheim.common.network.Message3d;
import net.minecraft.client.Minecraft;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;
import net.minecraft.world.storage.WorldInfo;

public class PacketHandlerClient {

	public static void handle(MessageParticles packet) {
		SpellEffectHandlerClient.select(Spells.values()[packet.i], packet.x, packet.y, packet.z, packet.x2, packet.y2, packet.z2);
	}

	public static void handle(MessageParty packet) {
		CardinalSystemClient.segment().party = packet.party;
		CardinalSystemClient.segment.partyIndex = 0;
	}

	public static void handle(MessageHotSpellC packet) {
		CardinalSystemClient.segment().hotSpells = packet.ids.clone();
	}

	public static void handle(MessageTileItem packet) {
		World world = Minecraft.getMinecraft().theWorld;
		TileEntity te = world.getTileEntity(packet.x, packet.y, packet.z);
		if (te != null && te instanceof ItemContainingTileEntity) ((ItemContainingTileEntity) te).setItem(packet.s);
	}

	public static void handle(MessageTimeStop packet) {
		if (packet.party == null) packet.party = new Party();
		TimeStopSystemClient.stop(packet.x, packet.y, packet.z, packet.party, packet.id);
	}

	public static void handle(Message1d packet) {
		switch (m1d.values()[packet.type]) {
			case DEATH_TIMER: AlfheimConfig.deathScreenAddTime = (int) packet.data1; break;
			case KNOWLEDGE: CardinalSystemClient.segment().knowledge[(int) packet.data1] = true; break;
			case TIME_STOP_REMOVE: TimeStopSystemClient.remove((int) packet.data1); break;
		}
	}

	public static void handle(Message2d packet) {
		switch (m2d.values()[packet.type]) {
			case ATTRIBUTE: {
				int d1 = (int) packet.data1;
				switch (d1) {
					case 0: EnumRace.setRaceID(Minecraft.getMinecraft().thePlayer, packet.data2); break;
					case 1: Flight.set(Minecraft.getMinecraft().thePlayer, packet.data2); break;
				}
			} break;
			case COOLDOWN: {
				SpellCastResult result = packet.data2 > 0 ? SpellCastResult.OK : SpellCastResult.values()[(int) -packet.data2];
				switch (result) {
					case DESYNC: throw new IllegalArgumentException("Client-server spells desynchronization. Not found spell for " + EnumRace.getByID(((int) packet.data1 >> 28) & 0xF) + " with id " + ((int) packet.data1 & 0xFFFFFFF));
					case NOMANA: ASJUtilities.say(Minecraft.getMinecraft().thePlayer, "alfheimmisc.cast.momana"); break; // TODO playSound "not enough mana"
					case NOTALLOW: ASJUtilities.say(Minecraft.getMinecraft().thePlayer, "alfheimmisc.cast.notallow"); break; // TODO playSound "not allowed"
					case NOTARGET: ASJUtilities.say(Minecraft.getMinecraft().thePlayer, "alfheimmisc.cast.notarget"); break; // TODO playSound "no target"
					case NOTREADY: /*ASJUtilities.say(Minecraft.getMinecraft().thePlayer, "alfheimmisc.cast.notready");*/ break; // TODO playSound "spell not ready"
					case NOTSEEING: ASJUtilities.say(Minecraft.getMinecraft().thePlayer, "alfheimmisc.cast.notseenig"); break; // TODO playSound "not seeing"
					case OBSTRUCT: ASJUtilities.say(Minecraft.getMinecraft().thePlayer, "alfheimmisc.cast.obstruct"); break; // TODO playSound "target obstructed"
					case OK: SpellCastingSystemClient.setCoolDown(AlfheimAPI.getSpellByIDs(((int) packet.data1 >> 28) & 0xF, (int) packet.data1 & 0xFFFFFFF), (int) packet.data2); break;
					case WRONGTGT: ASJUtilities.say(Minecraft.getMinecraft().thePlayer, "alfheimmisc.cast.wrongtgt"); break; // TODO playSound "wrong target"
				}
			} break;
			case UUID: CardinalSystemClient.segment().party.setUUID((int) packet.data2, (int) packet.data1); break;
			
		}
	}

	public static void handle(Message3d packet) {
		switch (m3d.values()[packet.type]) {
			case KEY_BIND:
				break;
			case PARTY_STATUS: {
				switch (PartyStatus.values()[(int) packet.data1]) {
					case DEAD: CardinalSystemClient.segment().party.setDead((int) packet.data2, ((int) packet.data3) == -10); break;
					case MANA: CardinalSystemClient.segment().party.setMana((int) packet.data2, (int) packet.data3); break;
				}
			} break;
			case WAETHER: {
				World world = Minecraft.getMinecraft().theWorld;
				WorldInfo info = Minecraft.getMinecraft().theWorld.getWorldInfo();
				info.setRaining(((int) packet.data1) > 0);
				info.setRainTime((int) packet.data2);
				info.setThundering(((int) packet.data1) > 1);
				info.setThunderTime((int) packet.data3);
			} break;
			case TOGGLER: ClientProxy.toggelModes(packet.data1 > 0, ((int) packet.data2 & 1) > 0, ((int) packet.data3 & 1) > 0, (((int) packet.data2 >> 1) & 1) > 0, (((int) packet.data3 >> 1) & 1) > 0); break;
		}
	}
}

File: src/main/java/alfheim/client/core/handler/PacketHandlerClient.kt
package alfheim.client.core.handler

import alexsocol.asjlib.*
import alexsocol.asjlib.extendables.block.TileItemContainer
import alfheim.api.AlfheimAPI
import alfheim.api.entity.*
import alfheim.api.spell.SpellBase.SpellCastResult
import alfheim.client.core.handler.CardinalSystemClient.PlayerSegmentClient
import alfheim.client.core.handler.CardinalSystemClient.SpellCastingSystemClient
import alfheim.client.core.handler.CardinalSystemClient.TimeStopSystemClient
import alfheim.client.core.proxy.ClientProxy
import alfheim.client.render.world.VisualEffectHandlerClient
import alfheim.client.render.world.VisualEffectHandlerClient.VisualEffects
import alfheim.common.core.handler.AlfheimConfigHandler
import alfheim.common.core.handler.CardinalSystem.KnowledgeSystem.Knowledge
import alfheim.common.core.handler.CardinalSystem.PartySystem.Party
import alfheim.common.core.handler.CardinalSystem.PartySystem.Party.PartyStatus
import alfheim.common.core.handler.ragnarok.RagnarokHandler
import alfheim.common.core.helper.*
import alfheim.common.entity.spell.EntitySpellFireball
import alfheim.common.item.relic.record.GinnungagapHandler
import alfheim.common.network.*
import alfheim.common.network.Message0dC.m0dc
import alfheim.common.network.Message1d.m1d
import alfheim.common.network.Message1l.m1l
import alfheim.common.network.Message2d.m2d
import alfheim.common.network.Message3d.m3d
import alfheim.common.network.MessageNI.mni
import net.minecraft.entity.EntityLivingBase
import net.minecraft.util.*

object PacketHandlerClient {
	
	fun handle(packet: MessageVisualEffect) {
		VisualEffectHandlerClient.select(VisualEffects.values()[packet.type], packet.data)
	}
	
	fun handle(packet: MessageParty) {
		PlayerSegmentClient.party = packet.party
		PlayerSegmentClient.partyIndex = 0
	}
	
	fun handle(packet: MessageHotSpellC) {
		PlayerSegmentClient.hotSpells = packet.ids.clone()
	}
	
	fun handle(packet: MessageTileItem) {
		val world = mc.theWorld
		val te = world.getTileEntity(packet.x, packet.y, packet.z)
		if (te is TileItemContainer) te.item = packet.s
	}
	
	fun handle(packet: MessageTimeStop) {
		if (packet.party == null) packet.party = Party()
		TimeStopSystemClient.stop(packet.x, packet.y, packet.z, packet.party!!, packet.id)
	}
	
	fun handle(packet: Message1d) {
		when (m1d.values()[packet.type]) {
			m1d.ESMABIL          -> PlayerSegmentClient.esmAbility = packet.data1 != 0.0
			m1d.DEATH_TIMER      -> AlfheimConfigHandler.deathScreenAddTime = packet.data1.I
			
			m1d.ELVEN_FLIGHT_MAX -> {
				AlfheimConfigHandler.flightTime = packet.data1.I
				ElvenFlightHelper.max = packet.data1
			}
			
			m1d.KNOWLEDGE        -> PlayerSegmentClient.knowledge.add("${Knowledge.values()[packet.data1.I]}")
			m1d.TIME_STOP_REMOVE -> TimeStopSystemClient.remove(packet.data1.I)
			m1d.GINNUNGAGAP      -> GinnungagapHandler.active = packet.data1 != 0.0
			
			m1d.RAGNAROK         -> {
				RagnarokHandler.ragnarok = packet.data1 < 1
				RagnarokHandler.fogFade = packet.data1.F
				
				if (0 < packet.data1 && packet.data1 < 1)
					mc.theWorld.playSound(mc.thePlayer.posX, mc.thePlayer.posY, mc.thePlayer.posZ, "mob.enderdragon.end", 50f, 0.5f, false)
			}
		}
	}
	
	fun handle(packet: Message1l) {
		when (m1l.values()[packet.type]) {
			m1l.SEED -> mc.theWorld.worldInfo.randomSeed = packet.data1
		}
	}
	
	fun handle(packet: Message2d) {
		when (m2d.values()[packet.type]) {
			m2d.ATTRIBUTE    -> {
				when (packet.data1.I) {
					0 -> mc.thePlayer.raceID = packet.data2.I
					1 -> mc.thePlayer.flight = packet.data2
				}
			}
			
			m2d.COOLDOWN     -> {
				when (if (packet.data2 > 0) SpellCastResult.OK else SpellCastResult.values()[(-packet.data2).I]) {
					SpellCastResult.DESYNC    -> throw IllegalArgumentException("Client-server spells desynchronization. Not found spell for ${EnumRace[packet.data1.I shr 28 and 0xF]} with id ${packet.data1.I and 0xFFFFFFF}")
					SpellCastResult.NOMANA    -> ASJUtilities.say(mc.thePlayer, "alfheimmisc.cast.momana")// TODO playSound "not enough mana"
					SpellCastResult.NOTALLOW  -> ASJUtilities.say(mc.thePlayer, "alfheimmisc.cast.notallow")// TODO playSound "not allowed"
					SpellCastResult.NOTARGET  -> ASJUtilities.say(mc.thePlayer, "alfheimmisc.cast.notarget")// TODO playSound "no target"
					SpellCastResult.NOTREADY  -> { /*ASJUtilities.say(mc.thePlayer, "alfheimmisc.cast.notready");*/
					}// TODO playSound "spell not ready"
					SpellCastResult.NOTSEEING -> ASJUtilities.say(mc.thePlayer, "alfheimmisc.cast.notseeing")// TODO playSound "not seeing"
					SpellCastResult.OBSTRUCT  -> ASJUtilities.say(mc.thePlayer, "alfheimmisc.cast.obstruct")// TODO playSound "target obstructed"
					SpellCastResult.OK        -> SpellCastingSystemClient.setCoolDown(AlfheimAPI.getSpellByIDs(packet.data1.I shr 28 and 0xF, packet.data1.I and 0xFFFFFFF), packet.data2.I)
					SpellCastResult.WRONGTGT  -> ASJUtilities.say(mc.thePlayer, "alfheimmisc.cast.wrongtgt")// TODO playSound "wrong target"
				}
			}
			
			m2d.UUID         -> PlayerSegmentClient.party?.setUUID(packet.data2.I, packet.data1.I)
			
			m2d.MODES        -> {
				if (packet.data1 > 0) ClientProxy.enableESM() else ClientProxy.disableESM()
				if (packet.data2 > 0) ClientProxy.enableMMO() else ClientProxy.disableMMO()
			}
			
			m2d.FIREBALLSYNC -> {
				(mc.theWorld.getEntityByID(packet.data1.I) as? EntitySpellFireball)?.target = mc.theWorld.getEntityByID(packet.data2.I) as? EntityLivingBase
			}
		}
	}
	
	fun handle(packet: Message3d) {
		when (m3d.values()[packet.type]) {
			m3d.KEY_BIND     -> Unit
			
			m3d.PARTY_STATUS -> {
				when (PartyStatus.values()[packet.data1.I]) {
					PartyStatus.DEAD      -> PlayerSegmentClient.party?.setDead(packet.data2.I, packet.data3.I == -10)
					PartyStatus.MANA      -> PlayerSegmentClient.party?.setMana(packet.data2.I, packet.data3.I)
					PartyStatus.HEALTH    -> PlayerSegmentClient.party?.setHealth(packet.data2.I, packet.data3.F)
					PartyStatus.MAXHEALTH -> PlayerSegmentClient.party?.setMaxHealth(packet.data2.I, packet.data3.F)
					PartyStatus.TYPE      -> PlayerSegmentClient.party?.setType(packet.data2.I, packet.data3.I)
				}
			}
			
			m3d.WAETHER      -> {
				mc.theWorld.setRainStrength(if (packet.data1.I > 0) 1f else 0f)
				mc.theWorld.setThunderStrength(if (packet.data1.I > 1) 1f else 0f)
				
				val info = mc.theWorld.worldInfo
				info.isRaining = packet.data1.I > 0
				info.rainTime = packet.data2.I
				info.isThundering = packet.data1.I > 1
				info.thunderTime = packet.data3.I
			}
			
			m3d.TOGGLER      -> ClientProxy.toggelModes(packet.data1 > 0, packet.data2.I and 1 > 0, packet.data3.I and 1 > 0, packet.data2.I shr 1 and 1 > 0, packet.data3.I shr 1 and 1 > 0)
		}
	}
	
	fun handle(packet: MessageNI) {
		when (mni.values()[packet.type]) {
			mni.WINGS_BL -> AlfheimConfigHandler.wingsBlackList = packet.intArray
		}
	}
	
	fun handle(packet: MessageSkinInfo) {
		CardinalSystemClient.playerSkinsData[packet.name] = packet.isFemale to packet.isSkinOn
	}
	
	fun handle(packet: Message0dC) {
		when (m0dc.values()[packet.type]) {
			m0dc.MTSPELL -> {
				val spell = AlfheimAPI.getSpellByIDs(KeyBindingHandlerClient.raceID, KeyBindingHandlerClient.spellID) ?: return
				mc.thePlayer?.addChatMessage(ChatComponentText(StatCollector.translateToLocalFormatted("spell.$spell.mtinfo", *spell.usableParams)))
			}
		}
	}
}